File: assignment.py
"""
    Copyright 2020 Lucas Bernard Black

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

class Assignment:
    # constructor
    def __init__(self, uid, title, due_date, completed=False, points=0.0, max_points=0.0, weight=0.0, link="", class_id=None):
        self.set_uuid(uid)
        self.set_title(title)
        self.set_due_date(due_date)
        self.set_completed(completed)
        self.set_points(points)
        self.set_max_points(max_points)
        self.set_weight(weight)
        self.set_link(link)
        self.set_class_id(class_id)

    def __str__(self):
        return self.title + " " + self.due_date

    def to_json(self):  # New special method
        grade = 0
        if self.get_max_points() > 0:
            grade = self.get_points() / self.get_max_points()
        return [ self.get_class_id(), self.get_due_date(), grade, self.get_weight()]

    # getters
    def get_uuid(self):
        return int(self.uid)

    def get_title(self):
        return self.title

    def get_due_date(self):
        return self.due_date

    def get_completed(self):
        return self.completed

    def get_link(self):
        if self.link is None:
            return ""
        return self.link

    def get_points(self):
        return float(self.points)

    def get_max_points(self):
        return float(self.max_points)

    def get_weight(self):
        if self.weight is None:
            return 0.0
        return self.weight

    def get_class_id(self):
        return int(self.class_id)

    # setters
    def set_uuid(self, uid):
        self.uid       = int(uid)

    def set_title(self, title):
        self.title     = str(title)

    def set_due_date(self, due_date):
        self.due_date  = str(due_date)

    def set_completed(self, completed):
        self.completed = bool(completed)

    def set_link(self, link):
        self.link      = str(link)

    def set_points(self, points):
        self.points = points

    def set_max_points(self, max_points):
        self.max_points = max_points

    def set_weight(self, weight):
        self.weight = float(weight)

    def set_class_id(self, class_id):
        self.class_id  = int(class_id)

File: static/js/assignment.js
/*

  Copyright 2020 Lucas Bernard Black

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

*/

colors = [
  "#ff5252",
  "#29b6f6",
  "#81c784",
  "#ffeb3b",
  "#40c4ff",
  "#673ab7",
  "#d32f2f",
  "#00e676",
  "#ff9800",
  "#cddc39",
];

function scatterAssignment(json_data, classes) {
    var ctx = document.getElementById("scatterAssignment");
    ctx.height = window.innerHeight * .3;

    // Define the data
    var datasets = [];
    var months = {
        "Jan": 0,
        "Feb": 1,
        "Mar": 2,
        "Apr": 3,
        "May": 4,
        "Jun": 5,
        "Jul": 6,
        "Aug": 7,
        "Sep": 8,
        "Oct": 9,
        "Nov": 10,
        "Dec": 11
    }

    // inefficient but works for now
    for (var j = 0; j < classes.length; j++) {
        var tempset = {
            label: classes[j][0],
            data: [],
            borderColor: colors[j],
            baclgroundColor: colors[j]
        }
        for (var i = 0; i < json_data.length; i++) {
            if (classes[j][4] == json_data[i][0]) {
                tempset.data.push({
                    x: months[json_data[i][1].substring(0,3)] * 30 + Number(json_data[i][1].substring(4, 6)),
                    y: json_data[i][2] * 100
                });
            }
        }
        datasets.push(tempset);
    }

    var options = {
        responsive: true,
        maintainAspectRatio: false,
        scales: {
            yAxes: [{
                ticks: {
                    callback: function(label, index, labels) {
                        return label.toString() + "%";
                    }
                },
                scaleLabel: {
                    display: true,
                    labelString: 'Grade of assignment'
                },
            }],
            xAxes: [{
                scaleLabel: {
                    display: true,
                    labelString: 'Approximate day of the year'
                }
            }]
        },
        tooltips: {
          callbacks: {
            label: function(item, data) {
              return '(' + item.label.toString() + ', ' + item.value.toString() + '%)';
            }
          }
        }
    };

    new Chart(ctx, {
        type: 'scatter',
        data: {
            datasets: datasets
        },
        options: options
    });
};